fdbserver/masterserver.actor.cpp

```diff
@@ -110,6 +113,145 @@ struct MasterData : NonCopyable, ReferenceCounted<MasterData> {
 	}
 };
 
+static std::pair<KeyRangeRef, bool> findRange(CoalescedKeyRangeMap<int>& key_resolver,
+                                              Standalone<VectorRef<ResolverMoveRef>>& movedRanges,
+                                              int src,
+                                              int dest) {
+	auto ranges = key_resolver.ranges();
+	auto prev = ranges.begin();
+	auto it = ranges.begin();
+	++it;
+	if (it == ranges.end()) {
+		if (ranges.begin().value() != src ||
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(ranges.begin()->range(), dest)) !=
+		        movedRanges.end())
+			throw operation_failed();
+		return std::make_pair(ranges.begin().range(), true);
+	}
+
+	std::set<int> borders;
+	// If possible expand an existing boundary between the two resolvers
+	for (; it != ranges.end(); ++it) {
+		if (it->value() == src && prev->value() == dest &&
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(it->range(), dest)) ==
+		        movedRanges.end()) {
+			return std::make_pair(it->range(), true);
+		}
+		if (it->value() == dest && prev->value() == src &&
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(prev->range(), dest)) ==
+		        movedRanges.end()) {
+			return std::make_pair(prev->range(), false);
+		}
+		if (it->value() == dest)
+			borders.insert(prev->value());
+		if (prev->value() == dest)
+			borders.insert(it->value());
+		++prev;
+	}
+
+	prev = ranges.begin();
+	it = ranges.begin();
+	++it;
+	// If possible create a new boundry which doesn't exist yet
+	for (; it != ranges.end(); ++it) {
+		if (it->value() == src && !borders.count(prev->value()) &&
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(it->range(), dest)) ==
+		        movedRanges.end()) {
+			return std::make_pair(it->range(), true);
+		}
+		if (prev->value() == src && !borders.count(it->value()) &&
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(prev->range(), dest)) ==
+		        movedRanges.end()) {
+			return std::make_pair(prev->range(), false);
+		}
+		++prev;
+	}
+
+	it = ranges.begin();
+	for (; it != ranges.end(); ++it) {
+		if (it->value() == src &&
+		    std::find(movedRanges.begin(), movedRanges.end(), ResolverMoveRef(it->range(), dest)) ==
+		        movedRanges.end()) {
+			return std::make_pair(it->range(), true);
+		}
+	}
+	throw operation_failed(); // we are already attempting to move all of the data one resolver is assigned, so do not
+	                          // move anything
+}
+
+// Balance key ranges among resolvers so that their load are evenly distributed.
+ACTOR Future<Void> resolutionBalancing(Reference<MasterData> self) {
```

Comment:
Just a thought: would it be cleaner to move this logic into its own file (may be ResolutionBalancer.actor.cpp) and let the Sequencer (better to rename Master -> Sequencer) only be responsible for `versions` related operations. 

`MasterCore` would add `resolutionBalancer` actor to its `collection` and make sure its new actor lifecycle is tied to the Sequencer. 

Thoughts?

Reply:
This actor accesses `MasterData`, so we have to move that into another separate header file. I'm inclined not doing that.

Reply:
Ahh got it. Well we can get away splitting the MasterData into multiple structs:
1. MasterServer fields needs access outside masterserver.actor 
2. Fields private only to masterserver.actor. 

For instance resolutionBalancing needs access to only below fields. 
```
resolverChanges,
resolvers
resolverChangeVersion
Version.
```

Again, it is not a necessary update and can be done as part of future cleanup PRs if needed. 

Thanks

Reply:
Done. PTAL!